linked_list/linked_list.py

```diff
@@ -1,79 +1,144 @@
+                return True
+            current = current.next
+        return False
+  
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def length(self):
-        pass
-
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_at_index(self, index):
-        pass
-
-    # method that returns the value of the last node in the linked list
-    # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_last(self):
-        pass
-
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def add_last(self, value):
-        pass
+    # Time Complexity:  O(n)
+    # Space Complexity: O(1) 
+    def length(self):  
+        if self.head == None:
+            return 0
+        current = self.head
+        count = 0
+        while current !=None:
+            count  +=1
+            current = current.next
+        return count
+
+    # method returns value at a given index in the linked list
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    def get_at_index(self, index):  
+        # if list is empty
+        if not self.head:
+            return None
+        current = self.head
+        current_i = 0
+        while current and current_i < index:
+            if index != current_i:
+                current_i +=1
+                current = current.next
+        return current.value
+
+    # method returns value of last node in linked list
+    # returns None if linked list is empty
+    # Time Complexity: O(n) - Would be O(1) if I added tail
+    # Space Complexity: O(1)
+    def get_last(self):  
+        if self.head == None:
+            return None
+        current = self.head
+        while current.next != None:
+            current = current.next
+        return current.value
+
+    # method - inserts a given value as a new last node in the linked list
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    #  trying add last
+    def add_last(self, value):  
+        last_node = Node(value)
+        # if list is empty, add and node becomes head (can I call add first, if so how?)
+        if self.head == None:
+            # last_node = Node(value)
+            last_node.next = self.head
+            self.head = last_node
+        else:
+            current = self.head
+            while current.next != None:
+                current = current.next
+            current.next = last_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
-    def find_max(self):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def find_max(self): 

```

Comment:
👍 But how is the space complexity O(n)  Are you creating any new lists or data structures?